packages/bitcore-lib/lib/transaction/transaction.js

```diff
@@ -1232,31 +1236,33 @@ Transaction.prototype.removeInput = function(txId, outputIndex) {
  * @param {Array|String|PrivateKey} privateKey
  * @param {number} sigtype
  * @param {String} signingMethod - method used to sign - 'ecdsa' or 'schnorr'
+ * @param {Buffer|String} merkleRoot - merkle root for taproot signing
  * @return {Transaction} this, for chaining
  */
-Transaction.prototype.sign = function(privateKey, sigtype, signingMethod) {
+Transaction.prototype.sign = function(privateKey, sigtype, signingMethod, merkleRoot) {
   $.checkState(this.hasAllUtxoInfo(), 'Not all utxo information is available to sign the transaction.');
   if (Array.isArray(privateKey)) {
     for (const pk of privateKey) {
-      this.sign(pk, sigtype, signingMethod);
+      this.sign(pk, sigtype, signingMethod, merkleRoot);
     }
     return this;
   }
-  for (const signature of this.getSignatures(privateKey, sigtype, signingMethod)) {
+  for (const signature of this.getSignatures(privateKey, sigtype, signingMethod, merkleRoot)) {
     this.applySignature(signature, signingMethod);
   }
   return this;
 };
 
-Transaction.prototype.getSignatures = function(privKey, sigtype, signingMethod) {
+Transaction.prototype.getSignatures = function(privKey, sigtype, signingMethod, merkleRoot) {
+  if (typeof merkleRoot === 'string') {
+    merkleRoot = Buffer.from(merkleRoot, 'hex');
+  }
   privKey = new PrivateKey(privKey);
-  sigtype = sigtype || Signature.SIGHASH_ALL;
-  var transaction = this;
-  var results = [];
-  var hashData = Hash.sha256ripemd160(privKey.publicKey.toBuffer());
-  for (let index = 0; index < this.inputs.length; index++) {
-    const input = this.inputs[index];
-    for (const signature of input.getSignatures(transaction, privKey, index, sigtype, hashData, signingMethod)) {
+  const results = [];
+  const hashData = Hash.sha256ripemd160(privKey.publicKey.toBuffer());
+  for (let i = 0; i < this.inputs.length; i++) {
```

Comment:
Consider flatMap & renaming 'results'
`
const hashData = Hash.sha256ripemd160(privKey.publicKey.toBuffer());
const results = this.inputs.flatMap((input, i) =>
  input.getSignatures(this, privKey, i, sigtype, hashData, signingMethod, merkleRoot)
);
`

Reply:
I'm inclined to keep this as is. The flatMap usage looks clean, but in practice it's pretty common to want to step into getSignatures with a debugger and/or try to debug why a specific input isn't returning expected sigs and the existing syntax is better for that.